docs/source/unreachable.rst

```diff
@@ -0,0 +1,147 @@
+.. _unreachable:
+
+********************************************
+Unreachable Code and Exhaustiveness Checking
+********************************************
+
+Sometimes it is necessary to write code that should never execute, and
+sometimes we write code that we expect to execute, but that is actually
+unreachable. The type checker can help in both cases.
+
+In this guide, we'll cover:
+
+- ``Never``, the primitive type used for unreachable code
+- ``assert_never()``, a helper for exhaustiveness checking
+- Directly marking code as unreachable
+- Detecting unexpectedly unreachable code
+
+``Never`` and ``NoReturn``
+==========================
+
+Type theory has a concept of a
```

Comment:
This is a really good introduction to this concept — one of the best descriptions I've seen. One other way that I've found useful in describing this concept is "an empty union". I don't know if adding that phrase here would help some people better understand the concept or add confusion. I mention it in case you think it might be useful.

Reply:
Thank you! I'll see if I can work that in too.

Reply:
Added